Add CI

tomleb opened the pull request:
## Issue: https://github.com/rancher/rancher/issues/44970

## Problem & Solution

Couple of things:
- This repository is not running any tests when creating PRs. Let's make it do that.
-  I'm keeping fossa as is because it IS working. Let's remove it once we actually start the migration.
- I'm not running golangci-lint because there are non-trivial lint errors to fix that would require its own review.

maxsokolovsky reviewed (dismissed):
Just curious: how do we know what version of Go CI would choose? Is it determined by `setup-go` and its hash?

tomleb commented:
> Just curious: how do we know what version of Go CI would choose? Is it determined by `setup-go` and its hash?

So in the GHA file, I'm setting:

```yaml
      with:
        go-version-file: 'go.mod'
```

which means it will look at the version in the `go.mod` file to determine which Go version to pick which [in this case is 1.22](https://github.com/rancher/dynamiclistener/blob/master/go.mod#L3). I'm not 100% sure where it will download that Go version though, but we officially trust this github action so no concern here.

tomleb commented:
I have removed the `Makefile` and the PR template for now to keep this PR simple. Since this is a repo who other contribute to, it's been suggested to seek feedback for adding a makefile.

I've changed the workflow so that it on runs on PR as well as when we push to new branches, but only for release branches. Seems to be what others are doing (eg: https://github.com/rancher/kube-api-auth/pull/23/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR33-R36, https://github.com/rancher/fleet/blob/main/.github/workflows/ci.yml#L9)
